### DISCUSSION TITLE
Long press event is not firing when clicking on left/right button

### DISCUSSION THREAD

**Discussion by zachariast** [react-native-tvos/react-native-tvos#409] [Q&A] [answered]
### Description

You guys rock!!

I am facing an issue(maybe a feature?) and I am not able to see an event when long pressing on left and right buttons on the RCU. I am currently testing it on Android Tv. 

Also when listening to the events I am receiving back the following:

`{"dispatchConfig": {}, "eventKeyAction": 1, "eventType": "right", "tag": 1763, "target": 1763}`

I am missing the **body** property

```
export type HWEvent = {
    eventType: 'up' | 'down' | 'right' | 'left' | 'blur' | 'focus' | 'pan' | string;
    eventKeyAction?: -1 | 1 | 0 | number;
    tag?: number;
    body?: {
      state: "Began" | "Changed" | "Ended",
      x: number,
      y: number,
      velocityx: number,
      velocityy: number
    }
  };
```

### Version

0.69.5-1

### Output of `npx react-native info`

```
System:
    OS: macOS 12.6.1
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
    Memory: 582.20 MB / 16.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 16.13.0 - ~/.nvm/versions/node/v16.13.0/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v16.13.0/bin/yarn
    npm: 8.1.0 - ~/.nvm/versions/node/v16.13.0/bin/npm
    Watchman: 2022.10.31.00 - /usr/local/bin/watchman
  Managers:
    CocoaPods: 1.11.3 - /usr/local/bin/pod
  SDKs:
    iOS SDK:
      Platforms: DriverKit 22.1, iOS 16.1, macOS 13.0, tvOS 16.1, watchOS 9.1
    Android SDK: Not Found
  IDEs:
    Android Studio: 2021.3 AI-213.7172.25.2113.9014738
    Xcode: 14.1/14B47b - /usr/bin/xcodebuild
  Languages:
    Java: 17.0.4 - /usr/local/opt/openjdk@17/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: 18.0.0 => 18.0.0
    react-native: Not Found
    react-native-macos: Not Found
    react-native-tvos:  0.69.5-1
  npmGlobalPackages:
    *react-native*: Not Found
```

### Steps to reproduce

I am listening for the events here: 

```
if (evt && evt.eventType === 'right') {
    if (evt.eventKeyAction === 1) {
        this.seekForward();
    }
} else if (evt && evt.eventType === 'left') {
    if (evt.eventKeyAction === 1) {
        this.seekBackward();
    }
}
```

**Comment by douglowder** [comment]
The native implementation on Android TV is in https://github.com/react-native-tvos/react-native-tvos/blob/tvos-v0.69.6/ReactAndroid/src/main/java/com/facebook/react/modules/core/ReactAndroidHWInputDeviceHelper.java .

We don't fill in the optional `body` property for TV remote events, with the exception of pan events on the touchpad found on the Siri remote of the Apple TV.

We do fire events for long presses of the center ("select") button -- the event type is "longSelect". We don't send long press events for the arrow keys. You should see events from the arrow keys for normal (not long) presses.

Normally we only send key action = ACTION_UP (1) so you don't need to check for the key action.

**Reply by zachariast** [reply]
So I raised the issue because I am trying to detect when the user long press the left/right arrows to seek in a video.
For example, if the user presses the right key on the RCU just once, we will fast forward the video by 10 seconds, but if he long presses the right arrow key, we should let the user seek around the video.

Do you have any recommendations on how to detect those long presses for the left/right arrows?

**Reply by douglowder** [reply]
@zachariast this would require adding native support for long presses of those keys, in the Android source file above.

There is an existing issue for making long press of the select button work better (https://github.com/react-native-tvos/react-native-tvos/issues/358)... When I pick that up, I will attempt to get long presses working for the arrow keys as well.

**Reply by esegebart** [reply]
This is exciting to hear! This is exactly what I was looking for, only I am trying to do something for Firestick (but is an android build). I am also trying to get seek to work by letting the user skip around the video since I have insanely long videos for the user to get through. I hope the long presses on the right and left arrows come soon! Is there anywhere I can add subscribe to wait for it to come out?

**Reply by esegebart** [reply]
@douglowder Could you tell me how to get the longSelect be recognized as an event? Will the longSelect be recognized using the enter key on a keyboard or only on the remote?

**Reply by douglowder** [reply]
longSelect should already be recognized either from the remote or from the keyboard. Arrow key long press support was added in 1f1f338698f2c864c6d7c5b7169c1ed48e44f119 and is already available in the latest release. You can now see `longLeft`, `longRight`, etc.

**Reply by danychi** [reply]
@esegebart 
for the long press in the arrows, you can detect it like this:
```ts
const myTVEventHandler = useCallback(
    (event: HWEvent) => {
        switch (event.eventType) {
            case 'longUp':
                handleLongUp();
            case 'longRight':
                handleLongRight();
                break;
            case 'longLeft':
                handleLongLeft();
                break;
            case 'longDown':
                handleLongDown();
                break;
        }
    },
    [],
);

useTVEventHandler(myTVEventHandler);
```

**Reply by esegebart** [reply]
@douglowder Thank you! This is great!

**Reply by esegebart** [reply]
@douglowder But longLeft/right is not recognizable from the remote? I held enter on the keyboard and 'select' printed out, but longSelect didn't.  It just acted like it was repeatedly pushing the button when I held it.  I must be doing something wrong.

Do I need to import HWEvent somewhere?

**Comment by mansilaad** [comment]
+1 same issue

**Comment by YungTosti** [comment]
Same issue here :(

**Comment by danychi** [comment]
I'm also facing the same issue, a pretty much-needed feature 😅

**Comment by douglowder** [comment, marked as answer]
longSelect should already be recognized either from the remote or from the keyboard. Arrow key long press support was added in 1f1f338698f2c864c6d7c5b7169c1ed48e44f119 (thanks to @danychi !) and is already available in the latest release. You can now see `longLeft`, `longRight`, etc.

**Comment by douglowder** [comment]
The next point release (0.69.6-4) will have this fix that makes `longSelect`, `longLeft`, etc., behave correctly and begin firing events while the button is still pressed, instead of waiting for the button to be released. https://github.com/react-native-tvos/react-native-tvos/pull/468

**Reply by esegebart** [reply]
Fantastic! Thanks Doug! Do you have any estimates on when that's hoping to be out? :)

**Reply by douglowder** [reply]
Within the next few days :)

**Reply by douglowder** [reply]
@esegebart please try this release candidate to make sure you are seeing the events you expect: https://www.npmjs.com/package/react-native-tvos/v/0.69.8-0rc3

**Comment by esegebart** [comment]
Fantastic thank you Doug!

**Comment by janvandertaelen** [comment]
@douglowder now it will indeed fire multiple longLeft, longRight & longPress but when you keep pressing the button there is no way now to know when you released a long press on a button. For instance currently working on a scrubber when you do a longPress we also need to know when the user released this button. Without it we cannot do the seek when the user releases the button. it just keeps spawning the event right now.

**Comment by douglowder** [comment]
@janvandertaelen you may want to turn on `ReactFeatureFlags.enableKeyDownEvents`. This will send a separate key up and key down event for each normal button press, and should send a key up at the end of the long press.

**Comment by janvandertaelen** [comment]
@douglowder for Android i can indeed check the eventKeyAction but for iOS it is missing ?

`{"dispatchConfig": {}, "eventType": "longRight"}`

**Reply by douglowder** [reply]
tvOS does not send key up or key down, it uses a native gesture recognizer to distinguish between long and short presses and send the correct event type to the JS layer.
